File: Data_Structures/fenwick_minmax.cpp
#include <bits/stdc++.h>
#include <limits>
using namespace std;

template<typename T, bool isMin>
struct Fenwick {
    int n;
    vector<T> fw;
    T id;

    Fenwick(int _n) {
        n = _n + 1;
        fw.assign(n, isMin ? numeric_limits<T>::max() : numeric_limits<T>::min());
        id = isMin ? numeric_limits<T>::max() : numeric_limits<T>::min();
    }

    static int lsb(int i) { return i & -i; }


    void update(int i, T val) {
        for (++i; i < n; i += lsb(i)) {
            if constexpr (isMin) fw[i] = min(fw[i], val);
            else fw[i] = max(fw[i], val);
        }
    }

    T query(int i) const {
        T res = id;
        for (++i; i > 0; i -= lsb(i)) {
            if constexpr (isMin) res = min(res, fw[i]);
            else res = max(res, fw[i]);
        }
        return res;
    }

    T range(int l, int r) const {
        return query(r) ^ query(l-1);
    }
};

File: datastructures/gol/fenwick/fenwick_minmax.go
package fenwick

import "math"

type FenwickMin struct {
	n  int
	fw []int
	id int
}

func NewFenwickMin(n int) *FenwickMin {
	fw := make([]int, n+1)
	for i := range fw {
		fw[i] = math.MaxInt
	}
	return &FenwickMin{n: n + 1, fw: fw, id: math.MaxInt}
}

func (f *FenwickMin) lsb(i int) int { return i & -i }

func (f *FenwickMin) Update(i, val int) {
	for i++; i < f.n; i += f.lsb(i) {
		f.fw[i] = min(f.fw[i], val)
	}
}

func (f *FenwickMin) Query(i int) int {
	res := f.id
	for i++; i > 0; i -= f.lsb(i) {
		res = min(res, f.fw[i])
	}
	return res
}

type FenwickMax struct {
	n  int
	fw []int
	id int
}

func NewFenwickMax(n int) *FenwickMax {
	fw := make([]int, n+1)
	for i := range fw {
		fw[i] = math.MinInt
	}
	return &FenwickMax{n: n + 1, fw: fw, id: math.MinInt}
}

func (f *FenwickMax) lsb(i int) int { return i & -i }

func (f *FenwickMax) Update(i, val int) {
	for i++; i < f.n; i += f.lsb(i) {
		f.fw[i] = max(f.fw[i], val)
	}
}

func (f *FenwickMax) Query(i int) int {
	res := f.id
	for i++; i > 0; i -= f.lsb(i) {
		res = max(res, f.fw[i])
	}
	return res
}
